engine/apps/integrations/mixins/alert_channel_defining_mixin.py

```diff
@@ -29,83 +31,92 @@ class AlertChannelDefiningMixin(object):
-                try:
-                    alert_receive_channel = AlertReceiveChannel.objects_with_deleted.get(
-                        token=kwargs["alert_channel_key"]
-                    )
-                except AlertReceiveChannel.DoesNotExist:
-                    cache.set(cache_key_short_term, CHANNEL_DOES_NOT_EXIST_PLACEHOLDER, self.CACHE_SHORT_TERM_TIMEOUT)
-                else:
-                    # Update short term cache
-                    serialized = serializers.serialize("json", [alert_receive_channel])
-                    cache.set(cache_key_short_term, serialized, self.CACHE_SHORT_TERM_TIMEOUT)
-
-                    # Update cached channels
-                    if cache.get(self.CACHE_DB_FALLBACK_OBSOLETE_KEY) is None:
-                        cache.set(self.CACHE_DB_FALLBACK_OBSOLETE_KEY, True, self.CACHE_DB_FALLBACK_REFRESH_INTERVAL)
-                        self.update_alert_receive_channel_cache()
-        except OperationalError:
-            logger.info("Cannot connect to database, using cache to consume alerts!")
+        alert_receive_channel, status = self.get_alert_receive_channel_from_cache(token)
 
-            # Searching for a channel in a cache
-            if cache.get(self.CACHE_KEY_DB_FALLBACK):
-                for obj in serializers.deserialize("json", cache.get(self.CACHE_KEY_DB_FALLBACK)):
-                    if obj.object.token == kwargs["alert_channel_key"]:
-                        alert_receive_channel = obj.object
+        if not alert_receive_channel or status == CHANNEL_DOES_NOT_EXIST_PLACEHOLDER:
+            logger.info(f"Integration {token} does not exist")
+            raise PermissionDenied(INTEGRATION_PERMISSION_DENIED_MESSAGE)
 
-                if alert_receive_channel is None:
-                    logger.info(f"Integration {kwargs['alert_channel_key']} not found in fallback cache")
-                    raise PermissionDenied(INTEGRATION_PERMISSION_DENIED_MESSAGE)
+        del kwargs["alert_channel_key"]
+        request = args[0]
+        request.alert_receive_channel = alert_receive_channel
+        finish = perf_counter()
+        logger.info(f"AlertChannelDefiningMixin finished in {finish - start}")
+        return super(AlertChannelDefiningMixin, self).dispatch(*args, **kwargs)
 
-            else:
-                logger.info("Cache is empty!")
-                raise
-        else:
-            if not alert_receive_channel:
-                logger.info(f"Integration {kwargs['alert_channel_key']} does not exist")
-                raise PermissionDenied(INTEGRATION_PERMISSION_DENIED_MESSAGE)
-            if alert_receive_channel.organization.is_moved:
+    def get_alert_receive_channel_from_cache(self, token: str) -> tuple[Optional[AlertReceiveChannel], Optional[str]]:
+        # Trying to define from short-term cache
+        cache_key_short_term = self.CACHE_KEY_SHORT_TERM + "_" + token
+        cached_alert_receive_channel_raw = cache.get(cache_key_short_term)
```

Comment:
curious why we use two different cache keys (this + `cache.get(self.CACHE_KEY_DB_FALLBACK)` down in `get_alert_receive_channel_from_db`)

Reply:
There are two usages of the cache, 1 which is short lived to optimize regular request traffic by caching individual channels and the other caches everything for if the DB goes down.  The keys are different because the fallback cache is initially loaded by StartupProbeView which does not have a specific channel to operate on.  The fallback channels never get expired (timeout=None) so they should always be available the other key is used to track expiry and update the channels by replacing the value.

Reply:
that makes sense, thanks for the breakdown 👍